router/internal/httpclient/traceclient.go

```diff
@@ -0,0 +1,147 @@
+package httpclient
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"net/http/httptrace"
+	"sync"
+	"time"
+)
+
+type ClientTraceContextKey struct{}
+
+type TraceInjectingRoundTripper struct {
+	base http.RoundTripper
+}
+
+func NewTraceInjectingRoundTripper(base http.RoundTripper) *TraceInjectingRoundTripper {
+	return &TraceInjectingRoundTripper{
+		base: base,
+	}
+}
+
+func GetClientTraceFromContext(ctx context.Context) *ClientTrace {
+	value := ctx.Value(ClientTraceContextKey{})
+	// Return no-op context if the subgraph context key was never set
+	if value == nil {
+		return &ClientTrace{
+			DialStart: make([]SubgraphDialStart, 0),
+			DialDone:  make([]SubgraphDialDone, 0),
+		}
+	}
+	return value.(*ClientTrace)
+}
+
+func (t *TraceInjectingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	trace := t.getClientTrace(req.Context())
+
+	req = req.WithContext(httptrace.WithClientTrace(req.Context(), trace))
+	trip, err := t.base.RoundTrip(req)
+
+	return trip, err
+}
+
+func InitTraceContext(ctx context.Context) context.Context {
+	trace := &ClientTrace{
+		DialStart: make([]SubgraphDialStart, 0),
+		DialDone:  make([]SubgraphDialDone, 0),
+	}
+	ctx = context.WithValue(ctx, ClientTraceContextKey{}, trace)
+	return ctx
+}
+
+func (t *TraceInjectingRoundTripper) getClientTrace(ctx context.Context) *httptrace.ClientTrace {
+	eC := GetClientTraceFromContext(ctx)
+
+	var dialStartMu sync.Mutex
+	var dialDoneMu sync.Mutex
+
+	trace := &httptrace.ClientTrace{
+		GetConn: func(hostPort string) {
+			eC.ConnectionCreate = &CreateSubgraphConnection{
+				Time:     time.Now(),
+				HostPort: hostPort,
+			}
+		},
+		GotConn: func(info httptrace.GotConnInfo) {
+			eC.ConnectionAcquired = &AcquiredSubgraphConnection{
+				Time:     time.Now(),
+				Reused:   info.Reused,
+				WasIdle:  info.WasIdle,
+				IdleTime: info.IdleTime,
+			}
+		},
+		GotFirstResponseByte: func() {
+			eC.FirstByte = &SubgraphFirstByte{
+				Time: time.Now(),
+			}
+		},
+		DNSStart: func(dnsStartInfo httptrace.DNSStartInfo) {
+			eC.DNSStart = &SubgraphDNSStart{
+				Time: time.Now(),
+				Host: dnsStartInfo.Host,
+			}
+		},
+		DNSDone: func(dnsDoneInfo httptrace.DNSDoneInfo) {
+			eC.DNSDone = &SubgraphDNSDone{
+				Time:      time.Now(),
+				Coalesced: dnsDoneInfo.Coalesced,
+				Error:     dnsDoneInfo.Err,
+			}
+		},
+		ConnectStart: func(network, addr string) {
+			start := SubgraphDialStart{
+				Time:    time.Now(),
+				Network: network,
+				Address: addr,
+			}
+
+			dialStartMu.Lock()
+			defer dialStartMu.Unlock()
+			eC.DialStart = append(eC.DialStart, start)
```

Comment:
From my testing I verified that multiple connection starts and done's can occur, I believe its important to record them all. Since we don't want a data race I use a lock.